theano/gpuarray/dnn.py

```diff
@@ -2800,6 +2800,204 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    _cop_num_outputs = 2
+    _f16_ok = True
+    default_output = 0
+
+    def __init__(self, dtype):
+        DnnBase.__init__(self, ["c_code/dnn_sptf.c"], "dnn_sptf")
+        self.dtype = dtype
+
+    def make_node(self, img, theta, grid_dims, desc, alpha=None, beta=None):
+        assert theta.dtype in ('float16', 'float32', 'float64')
+
+        context_name = infer_context_name(img)
+
+        theta = gpu_contiguous(as_gpuarray_variable(theta, context_name))
+        img = as_gpuarray_variable(img, context_name)
+        grid_dims = as_tensor_variable(grid_dims)
+
+        output = GpuArrayType(dtype=self.dtype,
+                              broadcastable=img.type.ndim * (False,),
+                              context_name=context_name)()
+        grid = GpuArrayType(dtype=self.dtype,
+                            broadcastable=img.type.ndim * (False,),
+                            context_name=context_name)()
+
+        if img.type.ndim != 4:
+            raise TypeError('img must be a 4D tensor')
+
+        if (not isinstance(desc.type, CDataType) or
+                desc.type.ctype != 'cudnnSpatialTransformerDescriptor_t'):
+            raise ValueError('desc must be cudnnSpatialTransformerDescriptor_t')
+
+        alpha = ensure_dt(alpha, _one, 'alpha', img.dtype)
+        beta = ensure_dt(beta, _zero, 'beta', img.dtype)
+
+        inputs = [img, theta, grid_dims, desc, alpha, beta]
+        outputs = [output, grid]
+        return Apply(self, inputs, outputs)
+
+    def L_op(self, inputs, outputs, grads):
+        img, theta, grid_dims, desc, alpha, beta = inputs
+
+        _, grid = outputs
+        dy = grads[0]
+
+        gop = GpuDnnTransformerGrad(self.dtype)(img, theta, grid, grid_dims, dy,
+                                                desc, alpha, beta)
+
+        return gop
+
+
+class GpuDnnTransformerGrad(DnnBase):
+    __props__ = ('dtype',)
+    _cop_num_inputs = 8
+    _cop_num_outputs = 2
+    _f16_ok = True
+
+    def __init__(self, dtype=theano.config.floatX):
+        DnnBase.__init__(self, ["c_code/dnn_sptf_grad.c"], "dnn_sptf_grad")
+        self.dtype = dtype
+
+    def make_node(self, img, theta, grid, grid_dims, dy, desc, alpha, beta):
+        context_name = infer_context_name(img)
+
+        dimg = GpuArrayType(dtype=self.dtype,
+                            broadcastable=img.type.ndim * (False,),
+                            context_name=context_name)()
+        dtheta = GpuArrayType(dtype=self.dtype,
+                              broadcastable=theta.type.ndim * (False,),
+                              context_name=context_name)()
+
+        inputs = [img, theta, grid, grid_dims, dy, desc, alpha, beta]
+        outputs = [dimg, dtheta,
+                   theano.gradient.grad_undefined(self, 2, grid_dims),
+                   theano.gradient.grad_undefined(self, 3, desc),
+                   theano.gradient.grad_undefined(self, 4, alpha),
+                   theano.gradient.grad_undefined(self, 5, beta)]
```

Comment:
I think at the end you may have 4 classes:
- GpuDnnTransformerDescriptor
- GpuDnnTransformer
- GpuDnnTransformerGradI
- GpuDnnTransformerGradW